### PR TITLE
updated code

### PR DESCRIPTION
**Pull request by lpullela**
this pr incorporates changes from the code review meeting including ensuring the optimizer updates between d and g are mutually exclusive, small change to extraction of image and language embeddings to get more data, and implement iterative training (perform a d and g step on the same data sample)

the new optimizer was already pushed to main in this pr: 
https://github.com/sophiamirrashidi/LLaVAFork/commit/cd499f559df007774f6f99e21b1e63b06577cf9c

Testing (make sure the updates help the model)

Training for one epoch

image

image

image

### CODE REVIEW

File: llava/VLLMSafety/discriminator.py [hunk 1]
```diff
@@ -6,58 +6,90 @@
...
 
-        img_tok = img_tok.view(-1, 5120) # image tokens have dim=3
+        for image, language in zipped_lists:
+            if d_mode:
+                loss, img, lang = self.forward(
+                    image.view(-1, 5120), language.view(-1, 5120), d_mode
+                )
+                total_loss += loss
+                img_is_correct += torch.sum(img)
+                lang_is_correct += torch.sum(lang)
```
Comment (lpullela) [inline llava/VLLMSafety/discriminator.py:41]: should we update the discriminator stochastically or in batches?

File: llava/VLLMSafety/discriminator.py [hunk 2]
```diff
@@ -6,58 +6,95 @@
...
         return x
- 
-    def forward(self, data, d_mode):
-        device = 'cuda'  
-        loss_function = nn.BCELoss() # from DCGAN
 
-        img_tok = data["image"]
-        lang_tok = data["lang"]
+    def run_forward(self, data, d_mode):
+
```
Comment (lpullela) [inline llava/VLLMSafety/discriminator.py:28]: can add some logic here to ensure that we send the same number of language and image tokens too to balance the set (typically more image data than lang)

File: llava/model/llava_arch.py [hunk 3]
```diff
@@ -158,6 +160,7 @@ def prepare_inputs_labels_for_multimodal(
                 images = [x.unsqueeze(0) if x.ndim == 3 else x for x in images]
             concat_images = torch.cat([image for image in images], dim=0)
             image_features = self.encode_images(concat_images)
+            raw_image_features = image_features
```
Comment (lpullela) [inline llava/model/llava_arch.py:163]: so that the discriminator sees the rawest form of the image embeddings

File: llava/train/llava_trainer.py [hunk 4]
```diff
@@ -849,4 +794,84 @@ def _inner_training_loop(
         if self.neftune_noise_alpha is not None:
             self._deactivate_neftune(self.model)
 
-        return TrainOutput(self.state.global_step, train_loss, metrics)
\ No newline at end of file
+        return TrainOutput(self.state.global_step, train_loss, metrics)
+    
+    def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor, Any]], dmode: bool = False) -> torch.Tensor:
+
+        inputs['d_mode'] = dmode
```
Comment (lpullela) [inline llava/train/llava_trainer.py:801]: control which parameters we track loss on

File: llava/train/llava_trainer.py [hunk 5]
```diff
@@ -849,4 +794,84 @@ def _inner_training_loop(
...
+
+        if not dmode: 
+            for name, param in model.named_parameters(): 
+                if "discriminator" in name: 
+                    param.requires_grad = False
+        
+        if dmode: 
+            for name, param in model.named_parameters(): 
+                if "discriminator" in name: 
+                    param.requires_grad = True
```
Comment (lpullela) [inline llava/train/llava_trainer.py:816]: ensures mutually exclusive updates 

the optimizer was fixed and pushed on main in this pr: 
https://github.com/sophiamirrashidi/LLaVAFork/commit/cd499f559df007774f6f99e21b1e63b06577cf9c

File: llava/model/language_model/llava_llama.py [hunk 6]
```diff
@@ -51,17 +57,27 @@ def __init__(self, config):
...
+
+        self.discriminator = Discriminator(5120)  # hard coding in sizes for now
 
         # Initialize weights and apply final processing
         self.post_init()
 
     def get_model(self):
         return self.model
+    
+    def log_data_to_json(self, data, date='10-10', directory="output_logs"):
```
Comment (sophiamirrashidi) [inline llava/model/language_model/llava_llama.py:75]: delete this - not needed

File: llava/model/language_model/llava_llama.py [hunk 7]
```diff
@@ -110,16 +128,21 @@ def forward(
...
-                return_dict=return_dict
+                return_dict=return_dict,
             )
 
             d_loss = discrim_dict["loss"]
-                        
-            model_output.loss = d_loss # returning only discriminator loss
+
+            data = {"disc loss": d_loss.item()}
+            self.log_data_to_json(data, date='10-10-d_mode')
```
Comment (sophiamirrashidi) [inline llava/model/language_model/llava_llama.py:139]: delete

File: llava/train/llava_trainer.py [hunk 8]
```diff
@@ -314,41 +318,7 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
...
-    #     if self.control.should_log and self.state.global_step > self._globalstep_last_logged:
-    #         if is_torch_tpu_available():
-    #             xm.mark_step()
-
-    #         logs: Dict[str, float] = {}
-            
-    #     if isinstance(disc_loss, np.ndarray):
-    #         disc_loss = torch.tensor(disc_loss)
-    #     if isinstance(summed_loss, np.ndarray):
-    #         summed_loss = torch.tensor(summed_loss)
```
Comment (sophiamirrashidi) [inline llava/train/llava_trainer.py:327]: uncomment and rethink what we are logging.

Comment (sophiamirrashidi) [reply llava/train/llava_trainer.py:327]: tbd

File: llava/train/llava_trainer.py [hunk 9]
```diff
@@ -747,25 +706,11 @@ def _inner_training_loop(
...
-                    elif abs(self.state.epoch - 0.5) < 1e-4:
-                        print(f"Saving checkpoint at epoch {self.state.epoch}")
-                        self._save_checkpoint(model, trial=None)
-
-                    elif abs(self.state.epoch - 0.75) < 1e-4:
-                        print(f"Saving checkpoint at epoch {self.state.epoch}")
-                        self._save_checkpoint(model, trial=None) 
-
                     self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
             # self._maybe_log_save_evaluate(tr_loss, disc_loss, summed_loss, model, trial, epoch, ignore_keys_for_eval)
```
Comment (sophiamirrashidi) [inline llava/train/llava_trainer.py:710]: bring back if keeping the monkeypatched logging

File: llava/train/llava_trainer.py (hunk 4 above)
Comment (sophiamirrashidi) [inline llava/train/llava_trainer.py:801]: d_mode

File: llava/model/language_model/llava_llama.py [hunk 10]
```diff
@@ -130,13 +153,14 @@ def forward(
                 use_cache=use_cache,
                 output_attentions=output_attentions,
                 output_hidden_states=output_hidden_states,
-                return_dict=return_dict
+                return_dict=return_dict,
             )
             
             model_output.loss = 1.5 * model_output.loss + d_loss
```
Comment (sophiamirrashidi) [inline llava/model/language_model/llava_llama.py:164]: remove 1.5 - htis is a hyperparameter

**Comment by lpullela** [comment]
eval with gpt 4.0-mini
